### DISCUSSION TITLE
How do I implement AsyncWrite for enum?

### DISCUSSION THREAD

**Discussion by Faputa** [tokio-rs/tokio#3993] [Q&A] [answered]
I have an enum like this:
```rust
use tokio::{io::WriteHalf, net::{tcp::OwnedWriteHalf, TcpStream}};
use tokio_rustls::server::TlsStream;

#[derive(Debug)]
pub enum TcpWriter {
    TcpWriter(OwnedWriteHalf),
    TlsTcpWriter(WriteHalf<TlsStream<TcpStream>>),
}
```
I want to implement AsyncWrite for it:
```rust
use tokio::io::AsyncWrite;

impl AsyncWrite for TcpWriter {
    // TODO
}
```
 Is it possible? What should I do?

**Comment by Darksonn** [comment, marked as answer]
An enum like this is available as [`tokio_util::either::Either`][1]. You can find the source in [`tokio-util/src/either.rs`][2].

[1]: https://docs.rs/tokio-util/0.6/tokio_util/either/enum.Either.html
[2]: https://github.com/tokio-rs/tokio/blob/master/tokio-util/src/either.rs

**Reply by Faputa** [reply]
Thank you very much, I'll look into it.
